### PR TITLE
fix(fw-datepicker): Fix datepicker validation issues

### PR DESCRIPTION
**Pull request by srivalli-sivaramasankaran**
This PR contains fix for all single datepicker related validation issues only. Adding the [design doc](https://www.figma.com/file/wOQ5hy91oBx9h8KHy25zPT/Date-and-Time-picker-new-updates?node-id=0%3A1&t=FxPn306DnudfZxBE-0) and the [contradicting issue list](https://docs.google.com/document/d/1zP48-aZaA5HRm1-GE-J5NaiBsFF4FkQtMOvFRnTgm2g/edit) for reference.

The conflicting and validation scenarios are deployed in a netlify link for reference : https://monumental-hamster-acb120.netlify.app/

Date range picker & date time picker - Yet to be addressed.

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [x] My commits have standard messages as mentioned in [Contributing Guidelines](./../blob/next/CONTRIBUTING.md)

### CODE REVIEW

**Review by arvindanta — CHANGES_REQUESTED** [review]

File: packages/crayons-core/src/components/datepicker/datepicker.scss [hunk 1]
```diff
@@ -290,6 +290,18 @@
...
 
+.invalid-alert {
+  position: relative;
+  display: inline-block;
+  overflow: hidden;
+
+  fw-icon {
+    cursor: pointer;
+    margin-block-start: 4px;
+    margin-inline-start: 5px;
```
[the remark this answers is not in the document]
Comment (srivalli-sivaramasankaran) [reply packages/crayons-core/src/components/datepicker/datepicker.scss]: fixed.

File: packages/crayons-core/src/components/datepicker/datepicker.tsx [hunk 2]
```diff
@@ -237,14 +255,23 @@ export class Datepicker {
    */
   @Event() fwBlur: EventEmitter;
 
+  /**
+   *   Triggered when text is entered in  input box.
```
[the remark this answers is not in the document]
Comment (srivalli-sivaramasankaran) [reply packages/crayons-core/src/components/datepicker/datepicker.tsx]: Fixed.

File: packages/crayons-core/src/components/datepicker/datepicker.tsx [hunk 3]
```diff
@@ -850,24 +822,116 @@ export class Datepicker {
...
-    if (
-      (this.maxYear && Number(this.year) > this.maxYear) ||
-      (this.minYear && Number(this.year) < this.minYear)
-    ) {
-      if (this.value) {
+  }
+
+  setDateAndErrorState(checkDate = false) {
+    this.isDateInvalid =
+      checkDate && !this.isDateWithinMinMaxDate(parseISO(this.value).valueOf())
```
Comment (arvindanta) [inline packages/crayons-core/src/components/datepicker/datepicker.tsx:829]: I think we don't need a ternary operator here. `checkDate && ...` should anyway be returning a boolean value only.

**Review by arvindanta — COMMENTED** [review]
Screenshot 2023-03-27 at 8 01 45 PM

With Max date as '2021-07-31' and no value explicitly set

In the year dropdown, I see 2021 and followed by 2023 . 2022 is missing .

I see this pattern in other examples too.
@srivalli-sivaramasankaran

**Comment by arvindanta** [comment]
`showErrorOnInvalidDate set to false and Min date '2020-07-31' min year as '2021' and invalid value set as 2019-07-31`

In the above case, the value is not being shown in the textbox. Is this the expectation?. When `showErrorOnInvalidDate` set to `true` , we show the value in text box and show the error.

File: packages/crayons-core/src/components/datepicker/datepicker.tsx [hunk 4]
```diff
@@ -850,24 +822,116 @@ export class Datepicker {
...
+        this.selectedTime = this.timeValue;
       }
-      this.year = this.minYear;
-      this.month = 0;
-      this.selectedDay = '1';
     }
   }
 
+  processValueChange(val, emitChange = false) {
+    // throw error if not ISO format and not display format
```
[the remark this answers is not in the document]
Comment (srivalli-sivaramasankaran) [reply packages/crayons-core/src/components/datepicker/datepicker.tsx]: fixed

File: packages/crayons-core/src/components/datepicker/datepicker.tsx (hunk 3 above)
Comment (srivalli-sivaramasankaran) [reply packages/crayons-core/src/components/datepicker/datepicker.tsx]: fixed.

**Comment by srivalli-sivaramasankaran** [comment]
> Screenshot 2023-03-27 at 8 01 45 PM
> 
> With Max date as '2021-07-31' and no value explicitly set
> 
> In the year dropdown, I see 2021 and followed by 2023 . 2022 is missing .
> 
> I see this pattern in other examples too. @srivalli-sivaramasankaran

This behaviour is expected as the fw-select component expects the year value in the dropdown and has the attribute 'selected'. Since the invalid date value's year value will not be in the year dropdown, we are adding just the year value but not the years in between them, and removing once the user moves to a valid year.

**Review by arvindanta — APPROVED** [review]

**Pull request merged**